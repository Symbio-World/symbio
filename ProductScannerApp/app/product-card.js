import React from 'react';
import {
  Text,
} from 'react-native';
import {
  Button,
  Card,
  Icon,
} from 'react-native-elements'

export const ProductCard = ({
  image,
  name,
  children,
}) => (
  <Card
    title={name}
    image={{
      uri: image
    }}
  >
    <Text style={{ marginBottom: 10 }}>
      {children}
    </Text>
    <Button
      icon={<Icon name='code' color='#ffffff' />}
      buttonStyle={{ borderRadius: 0, marginLeft: 0, marginRight: 0, marginBottom: 0 }}
      title='VIEW NOW'
    />
  </Card>
)