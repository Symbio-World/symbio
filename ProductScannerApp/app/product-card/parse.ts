import cheerio from 'cheerio'
import { Parse } from 'fetcher-core'

// TODO to test
export const parse: Parse = (html) => {
  const $ = cheerio.load(html)

  return {
    ingredients: $("[id$='ingredients']").text(),
    allergens: $('#info')
      .find('table')
      .find('td')
      .eq(1)
      .text(),
    origin: $('#origin')
      .find('p')
      .text(),
  }
}
