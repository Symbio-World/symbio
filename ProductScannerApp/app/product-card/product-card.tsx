import React from 'react'
import { Image, Text, ScrollView } from 'react-native'
import { t } from 'react-native-tailwindcss'
import { ProductData } from 'fetcher-core'

type Props = Omit<ProductData, 'links'>

export const ProductCard: React.FC<Props> = props => (
  <ScrollView style={[t.bgWhite, t.flex1, t.p4]}>
    <Image style={[t.h40]} source={{ uri: props.image }} />
    <Text style={[t.textLg, t.textCenter]}>{props.name}</Text>
    <Text>{props.description}</Text>
    <Text>{props.ingredients}</Text>
    <Text>{props.allergens}</Text>
    <Text>{props.origin}</Text>
  </ScrollView>
)
