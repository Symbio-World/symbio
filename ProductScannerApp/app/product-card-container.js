import React from 'react';
import { ActivityIndicator, Text } from 'react-native'

import { fetchProduct } from './product-api'
import { ProductCard } from './product-card'

export class ProductCardContainer extends React.Component {
  state = {}

  componentDidMount() {
    this.fetchProduct()
  }

  async fetchProduct() {
    const product = await fetchProduct(this.props.barcode)
      .catch(error => {
        this.setState({
          error
        })
      })
    this.setState({ product })
  }

  render() {
    if (this.state.product) {
      return <ProductCard {...this.state.product} />
    }

    if (this.state.error) {
      return <Text>There was an error, please retry</Text>
    }

    return <ActivityIndicator size='large' color="#00ff00" />
  }
}