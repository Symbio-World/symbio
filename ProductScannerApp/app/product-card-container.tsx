import React, { useState, useEffect } from 'react'
import { ActivityIndicator, Text } from 'react-native'

import {
  fetchProductData,
  FetchProductData,
  ProductData,
} from './fetch-product-data'
import { ProductCard } from './product-card'

type Props = {
  barcode: string
}

type Deps = {
  fetchProductData: FetchProductData
}

type CreateProductCardContainer = (deps: Deps) => React.FC<Props>

export const createProductCardContainer: CreateProductCardContainer = ({
  fetchProductData,
}) => props => {
  const [product, setProduct] = useState<ProductData>(null)
  // const [error, setError] = useState<Error>(null)

  const fetchProduct = async () => {
    const product = await fetchProductData(props.barcode)
    // .catch(error => {
    //   this.setState({
    //     error
    //   })
    // })
    setProduct(product)
  }

  useEffect(() => {
    fetchProduct()
  }, [])

  if (product) {
    return <ProductCard {...product} />
  }

  // if (error) {
  //   return <Text>There was an error, please retry</Text>
  // }

  return <ActivityIndicator size="large" color="#00ff00" />
}

export const ProductCardContainer = createProductCardContainer({
  fetchProductData,
})
