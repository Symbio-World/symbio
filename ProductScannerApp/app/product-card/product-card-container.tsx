import React, { useState, useEffect } from 'react'
import { ActivityIndicator } from 'react-native'
import {
  FetchProductData,
  ProductData,
} from 'fetcher-core'
import { ProductCard } from './product-card'

type Props = {
  barcode: string
}

type Deps = {
  fetchProductData: FetchProductData
}

type CreateProductCardContainer = (deps: Deps) => React.FC<Props>

export const createProductCardContainer: CreateProductCardContainer = ({
  fetchProductData,
}) => props => {
  const [product, setProduct] = useState<ProductData>()
  // const [error, setError] = useState<Error>(null)

  const fetchProduct = async () => {
    const product = await fetchProductData(props.barcode)
    // .catch(error => {
    //   this.setState({
    //     error
    //   })
    // })
    setProduct(product)
  }

  useEffect(() => {
    fetchProduct()
  }, [])

  if (product) {
    return <ProductCard {...product} />
  }

  // if (error) {
  //   return <Text>There was an error, please retry</Text>
  // }

  return <ActivityIndicator size="large" color="#00ff00" />
}
