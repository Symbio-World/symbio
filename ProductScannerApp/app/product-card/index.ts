import axios from 'axios'
import cheerio from 'cheerio'
import {
  GOOGLE_CUSTOM_SEARCH_ENGINE_KEY,
  GOOGLE_CUSTOM_SEARCH_ENGINE_CX,
  GOOGLE_TRANSLATE_API_KEY,
} from 'react-native-dotenv'
import {
  createFetchProduct,
} from 'fetcher-core'

import { createProductCardContainer } from './product-card-container'

export const ProductCardContainer = createProductCardContainer({
  fetchProductData: createFetchProduct({
    fetch: axios.request,
    parse: (html) => {
      const $ = cheerio.load(html)

      return {
        ingredients: $("[id$='ingredients']").text(),
        allergens: $('#info')
          .find('table')
          .find('td')
          .eq(1)
          .text(),
        origin: $('#origin')
          .find('p')
          .text(),
      }
    },
    googleSeachConfig: {
      key: GOOGLE_CUSTOM_SEARCH_ENGINE_KEY,
      cx: GOOGLE_CUSTOM_SEARCH_ENGINE_CX,
      url: 'https://www.googleapis.com/customsearch/v1'
    },
    googleTranslateApiConfig: {
      key: GOOGLE_TRANSLATE_API_KEY,
      url: 'https://translation.googleapis.com/language/translate/v2',
      target: 'en'
    }
  }),
})
