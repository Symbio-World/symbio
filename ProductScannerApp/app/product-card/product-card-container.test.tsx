import 'react-native'
import React from 'react'
import { create, act, ReactTestRenderer } from 'react-test-renderer'
import { ActivityIndicator } from 'react-native'

import { createProductCardContainer } from './product-card-container'
import { ProductCard } from './product-card'

describe('ProductCardContainer', () => {
  it('renders correctly', () => {
    const fetchProductData = jest.fn<any, any>(() => Promise.resolve())
    const ProductCardContainer = createProductCardContainer({
      fetchProductData,
    })
    create(<ProductCardContainer barcode="6414893012318" />)
  })

  it('renders loading screen at the start', () => {
    const fetchProductData = jest.fn<any, any>(() => Promise.resolve())
    const ProductCardContainer = createProductCardContainer({
      fetchProductData,
    })
    const tree = create(
      <ProductCardContainer barcode="6414893012318" />,
    )
    const root = tree.root
    const activityIndicators = root.findAllByType(ActivityIndicator)
    expect(activityIndicators.length).toBe(1)
  })

  it('renders product card if data fetch succeeded', async () => {
    const promise = Promise.resolve({})
    const fetchProductData = jest.fn<any, any>(() => promise)
    const ProductCardContainer = createProductCardContainer({
      fetchProductData,
    })
    let tree: ReactTestRenderer
    act(() => {
      tree = create(
        <ProductCardContainer barcode={'6414893012318'} />,
      )
    })
    const root = tree!.root
    let activityIndicators = root.findAllByType(ActivityIndicator)
    expect(activityIndicators.length).toBe(1)

    await act(async () => {
      await promise
    })
    activityIndicators = root.findAllByType(ActivityIndicator)
    expect(activityIndicators.length).toBe(0)
    const productCards = root.findAllByType(ProductCard)
    expect(productCards.length).toBe(1)
  })

  // it('renders error card if error occurs', async () => {
  //   const promise = Promise.reject(new Error());
  //   (getProductData as jest.Mock).mockImplementationOnce(() => promise)
  //   const tree = renderer.create(<ProductCardContainer barcode={'6414893012318'} />)
  //   await Promise.resolve()
  //   const root = tree.root
  //   const activityIndicators = root.findAllByType(ActivityIndicator)
  //   expect(activityIndicators.length).toBe(0)
  //   const productCards = root.findAllByType(Text)
  //   expect(productCards.length).toBe(1)
  // })
})
