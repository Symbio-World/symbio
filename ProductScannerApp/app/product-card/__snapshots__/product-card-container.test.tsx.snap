// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCardContainer renders correctly 1`] = `
<ActivityIndicator
  animating={true}
  color="#00ff00"
  hidesWhenStopped={true}
  size="large"
/>
`;
